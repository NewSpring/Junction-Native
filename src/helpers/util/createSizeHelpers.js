//@flow

type ISizeList = [string, (number) => number][];
export const sizes: ISizeList = [
  ["none", () => 0],
  ["xs", x => x / 4],
  ["sm", x => x / 2],
  ["md", x => x],
  ["lg", x => x * 2],
  ["xl", x => x * 4],
];

export default (
  { place, style }: { place: string, style: Function },
  sizeList: ISizeList = sizes,
): { style: Function, name: string }[] =>
  sizeList.map(([size, modifier]) => ({
    style: style.map(x => ({ ...x, [place]: modifier(x[place]) })),
    name: size,
  }));
