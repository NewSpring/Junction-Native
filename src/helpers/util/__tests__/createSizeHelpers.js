import Style from "further";

import createSizeHelpers, { sizes } from "../createSizeHelpers";

it("has a default scale", () => {
  expect(sizes).toMatchSnapshot();
});

it("modifies a passed Style instance at the given place", () => {
  const st = Style.of({ margin: 10 });
  const mappedSizes = createSizeHelpers({ place: "margin", style: st });

  const [none, xs, sm, md, lg, xl] = mappedSizes;

  const check = ({ style, name }, result, str) => {
    expect(style.resolve()).toEqual({ margin: result });
    expect(name).toEqual(str);
  };

  check(none, 0, "none");
  check(xs, 10 / 4, "xs");
  check(sm, 5, "sm");
  check(md, 10, "md");
  check(lg, 20, "lg");
  check(xl, 40, "xl");
});

it("allows a custom size map (type independent)", () => {
  const st = Style.of({ margin: 10 });
  const customList = [["pancake", x => x + "pancake"]];
  const mappedSizes = createSizeHelpers(
    { place: "margin", style: st },
    customList,
  );

  const [pancake] = mappedSizes;

  const check = ({ style, name }, result, str) => {
    expect(style.resolve()).toEqual({ margin: result });
    expect(name).toEqual(str);
  };

  check(pancake, "10pancake", "pancake");
});
