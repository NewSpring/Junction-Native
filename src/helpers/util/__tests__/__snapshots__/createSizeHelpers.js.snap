// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has a default scale 1`] = `
Array [
  Array [
    "none",
    [Function],
  ],
  Array [
    "xs",
    [Function],
  ],
  Array [
    "sm",
    [Function],
  ],
  Array [
    "md",
    [Function],
  ],
  Array [
    "lg",
    [Function],
  ],
  Array [
    "xl",
    [Function],
  ],
]
`;
